[docs] Add migration guide for the picker's `ownerState` changes (#17151)
mui/mui-x: docs/data/migration/migration-pickers-v7/migration-pickers-v7.md

#### Migrate `slotProps.textField`

If you are passing props to `slotProps.textField`,

> these props are now received by `PickersTextField` and should keep working the same way as before.

Reviewer: ```suggestion
the `PickersTextField` component now receives these props and should keep working the same as before.
```

});
```

### Clean the `ownerState`

The `ownerState` is an object describing the current state of a given component to let you do more advanced customization.
It is used in two different APIs:

1. In the theme's `styleOverrides`, the `ownerState` is passed to the function that returns the style:

```ts
const theme = createTheme({
  components: {
    MuiDateCalendar: {
      styleOverrides: {
        root: ({ ownerState }) => ({
          /** Style based on the ownerState */
        }),
      },
    },
  },
});
```

2. In the `slotProps`, the `ownerState` is passed to the function that returns the custom props:

```tsx
<DatePicker
  slotProps={{
    actionBar: (ownerState) => ({
      /** Props based on the ownerState */
    }),
  }}
/>
```

> Before version `v8.x`, the `ownerState` contained every prop of the component that was defining it, plus some additional internal states if needed.

Reviewer: Do we need that specific clarification?
I don't see any additional clarity it brings... 🤔 🤷 
```suggestion
Before version `v8.x`, the `ownerState` contained every prop of the component, plus some additional internal states if needed.
```

> Before version `v8.x`, the `ownerState` was equal to the props of the component that was defining it, plus some additional internal states if needed.

Reviewer 2: ```suggestion
Before version `v8.x`, the `ownerState` contained every prop of the component that was defining it, plus some additional internal states if needed.
```

> This came with a few problems:

Reviewer: ```suggestion
This presented a few problems:
```

> - It is hard to know which parent defines the `ownerState` and therefore which props it contains (is the `actionBar` slot handled by `DatePicker`? by `DesktopDatePicker`? by `PickerLayout`?).

Reviewer 2: ```suggestion
- It is hard to know which ancestor defines the `ownerState` and therefore which props it contains (is the `actionBar` slot handled by `DatePicker`, by `DesktopDatePicker` or by `PickerLayout`?).
```
maybe?

> - It is hard to know which ancestor defines the `ownerState` and therefore which props it contains (is the `actionBar` slot handled by `DatePicker`, by `DesktopDatePicker` or by `PickerLayout`?).

Reviewer: ```suggestion
- It was hard to know which ancestor defines the `ownerState` and therefore which props it contains (is the `actionBar` slot handled by `DatePicker`, by `DesktopDatePicker` or by `PickerLayout`?).
```

> - Many properties of the `ownerState` were not meant for public usage, which made the evolution of the codebase complex because all the props received by an internal component became public API and any change introduced a breaking change.

Reviewer: ```suggestion
- Many properties of the `ownerState` were not intended for public use, which complicated the evolution of the codebase. All the props received by an internal component became part of the public API, and consequently, any changes to them would have resulted in a breaking change.
```

> - A lot of the properties of the `ownerState` were not meant for public usage, which made the evolution of the codebase complex because all the props received by an internal component becomes public API and changing them becomes a breaking change.

Reviewer 2: ```suggestion
- Many properties of the `ownerState` were not meant for public usage, which made the evolution of the codebase complex because all the props received by an internal component became public API and any change introduced a breaking change.
```

> - Some properties actually useful to customize a component were not present, if the component was not using them by default (if the built-in styles for the `actionBar` don't need to know if the picker is disabled, then the `ownerState` won't contain this information).

Reviewer 2: ```suggestion
- Some properties that would have been useful for customizing a component were not present, if the component was not using them by default. For example, if the built-in styles for the `actionBar` don't need to know if the picker is disabled, then the `ownerState`  of the `actionBar` won't contain this information.
```

> - Some properties that would have been useful for customizing a component were not present, if the component was not using them by default. For example, if the built-in styles for the `actionBar` don't need to know if the picker is disabled, then the `ownerState` of the `actionBar` won't contain this information.

Reviewer: ```suggestion
- Some properties that would have been useful for customizing a component were not present if the component was not using them by default. For example, if the built-in styles for the `actionBar` didn't need to know if the picker is disabled, then the `ownerState` of the `actionBar` didn't contain this information.
```

> - It is hard to understand which element some properties apply to. If the `ownerState` of the `monthButton` slot contains `disabled`, is it because the button is disabled or because the picker is disabled?

Reviewer 2: ```suggestion
- The naming of the props made it difficult to understand which element they apply to. If the `ownerState` of the `monthButton` slot contains `disabled`, it is hard to establish whether the disabled state applies to the `monthButton` or the picker itself.
```

> - The naming of the props made it difficult to understand which element they apply to. If the `ownerState` of the `monthButton` slot contains `disabled`, it is hard to establish whether the disabled state applies to the `monthButton` or the picker itself.

Reviewer: ```suggestion
- The naming of the props made it difficult to understand which element they applied to. If the `ownerState` of the `monthButton` slot contained `disabled`, it was hard to establish whether the disabled state applied to the `monthButton` or the Picker itself.
```

To solve these issues, the `ownerState` has been reworked.

> All the component's `ownerState` contain a shared set of properties describing the state of the picker they are in (`isPickerValueEmpty`, `isPickerOpen`, `isPickerDisabled`, `isPickerReadOnly`, `pickerVariant` and `pickerOrientation`).

Reviewer 2: ```suggestion
Every component's `ownerState` contains a shared set of properties describing the state of the picker it is in (`isPickerValueEmpty`, `isPickerOpen`, `isPickerDisabled`, `isPickerReadOnly`, `pickerVariant` and `pickerOrientation`).
```

> Some component's `ownerState` contain additional properties describing their own state (`isMonthDisabled` for the month button, `toolbarOrientation` for the toolbar, `isDaySelected` for the day button, ...).

Reviewer: ```suggestion
Some component's `ownerState` contain additional properties describing their own state (`isMonthDisabled` for the month button, `toolbarOrientation` for the toolbar, `isDaySelected` for the day button, etc.).
```
